### PR TITLE
8331736: C2: Live Node limit exceeded limit after JDK-8316991

### PR DESCRIPTION
**Pull request by JohnTortugo**
Please, consider this patch to interrupt execution of `split_unique_types` (SUT) when number of live nodes reaches 3/4 of `max_live_nodes`.

The included test case reproduces the problem. The number of live nodes before running phase 3 of SUT is ~20k, after processing about 750 mergemem nodes the number of live nodes is over 70k. This problem was first encountered when running an old `.jar` file that was created before `invokedynamic` optimizations - that's why I disable string optimizations in the test case. The test case is strongly based on [the method that was originally triggering the problem](https://github.com/Unidata/netcdf-java/blob/c782ef80ab54a09befd6d5065c6baeed54949222/cdm/radial/src/main/java/ucar/nc2/iosp/nids/Nidsheader.java#L2174).

Tested on Linux, Win, Mac x86_64 tier1-3 and GHA.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8331736](https://bugs.openjdk.org/browse/JDK-8331736): C2: Live Node limit exceeded limit after JDK-8316991 (**Bug** - P3)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [1e92268a](https://git.openjdk.org/jdk/pull/19545/files/1e92268a08306e7e93879ed20e86d884882729b8)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to [1e92268a](https://git.openjdk.org/jdk/pull/19545/files/1e92268a08306e7e93879ed20e86d884882729b8)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19545/head:pull/19545` \
`$ git checkout pull/19545`

Update a local copy of the PR: \
`$ git checkout pull/19545` \
`$ git pull https://git.openjdk.org/jdk.git pull/19545/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19545`

View PR using the GUI difftool: \
`$ git pr show -t 19545`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19545.diff">https://git.openjdk.org/jdk/pull/19545.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19545#issuecomment-2148222887)

### CODE REVIEW

**Review by TobiHartmann — APPROVED** [review]
Looks good to me. We have similar logic for split-if (see `must_throttle_split_if`).

File: test/hotspot/jtreg/compiler/c2/TestScalarReplacementMaxLiveNodes.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+ */
+
+/*
+ * @test
+ * @bug 8331736
+ * @summary Check that C2 does not exceed max live node limit when splitting unique types of large allocation merge.
+ * @library /test/lib /
+ * @requires vm.debug & vm.flagless & vm.compiler2.enabled & vm.opt.final.EliminateAllocations
+ * @compile -XDstringConcat=inline TestScalarReplacementMaxLiveNodes.java
+ * @run main/othervm -Xbatch -server -XX:-OptimizeStringConcat -XX:-TieredCompilation
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/c2/TestScalarReplacementMaxLiveNodes.java:31]:
```suggestion
 * @run main/othervm -Xbatch -XX:-OptimizeStringConcat -XX:-TieredCompilation
```

File: test/hotspot/jtreg/compiler/c2/TestScalarReplacementMaxLiveNodes.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8331736
+ * @summary Check that C2 does not exceed max live node limit when splitting unique types of large allocation merge.
+ * @library /test/lib /
+ * @requires vm.debug & vm.flagless & vm.compiler2.enabled & vm.opt.final.EliminateAllocations
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/c2/TestScalarReplacementMaxLiveNodes.java:29]:
```suggestion
 * @requires vm.debug & vm.compiler2.enabled
```

It think it's best to not overly restrict the test to get some more coverage from different flag combinations in the CI.

File: src/hotspot/share/opto/escape.cpp [hunk 3]
```diff
@@ -4793,6 +4793,20 @@ void ConnectionGraph::split_unique_types(GrowableArray<Node *>  &alloc_worklist,
         nmm->set_memory_at(ni, result);
       }
     }
+
+    // If we have crossed the 3/4 point of max node limit it's too risky
+    // to continue with EA/SR because we might hit the max node limit.
+    if (_compile->live_nodes() >= _compile->max_node_limit()*0.75) {
```
Comment (chhagedorn) [inline src/hotspot/share/opto/escape.cpp:4799]:
```suggestion
    if (_compile->live_nodes() >= _compile->max_node_limit() * 0.75) {
```

**Review by chhagedorn — APPROVED** [review]
Apart from Tobias' suggestions, it looks good to me, too!

**Comment by JohnTortugo** [comment]
Thank you @TobiHartmann, @chhagedorn for reviewing once more! I addressed your feedback.

**Pull request closed without merging**